src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java

```diff
@@ -1522,6 +1576,8 @@ private final void writeInteger(int value) {
         // Segment support.
         private boolean segmentMode;
         private boolean allowSegmented;
+        // Write data directly to underlying stream. Don't use internal buffer.
+        private boolean unbufferedMode;
```

Comment:
I would suggest replace `unbufferedMode` with `bufferedMode` field. It will simplify logic a little bit.
Then you will need to rename the method `fillSegmentSizeAndEnableUnbufferedMode` to be `fillSegmentSizeAndDisableBufferedMode`. 
There are several places where `!unbufferedMode` will be replaced with `bufferedMode` .